plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/CogImageInputStreamSpi.java

```diff
@@ -0,0 +1,82 @@
+/*
+ *    ImageI/O-Ext - OpenSource Java Image translation Library
+ *    http://www.geo-solutions.it/
+ *    http://java.net/projects/imageio-ext/
+ *    (C) 2019, GeoSolutions
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    either version 3 of the License, or (at your option) any later version.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package it.geosolutions.imageioimpl.plugins.cog;
+
+import javax.imageio.spi.ImageInputStreamSpi;
+import javax.imageio.spi.ServiceRegistry;
+import javax.imageio.stream.ImageInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.Locale;
+
+/**
+ * SPI for creating either `CachingCogImageInputStream` or `DefaultCogImageInputStream`.
+ *
+ * @author joshfix
+ * Created on 2019-08-23
+ */
+public class CogImageInputStreamSpi extends ImageInputStreamSpi {
+
+    private static final String vendorName = "GeoSolutions";
+    private static final String version = "1.0";
+    private static final Class<CogUri> inputClass = CogUri.class;
+
+    public CogImageInputStreamSpi() {
+        super(vendorName, version, inputClass);
+    }
+
+    @Override
+    public ImageInputStream createInputStreamInstance(Object input, boolean useCache, File cacheDir) throws IOException {
+        if (input instanceof CogUri) {
+            if (((CogUri)input).isUseCache()) {
+                return new CachingCogImageInputStream((CogUri) input);
+            } else {
+                return new DefaultCogImageInputStream((CogUri) input);
+            }
+        }
+        if (input instanceof String || input instanceof URL || input instanceof URI) {
+            if (((CogUri)input).isUseCache()) {
```

Comment:
This seems to guarantee an ClassCastException, given the input might only be a string, url or uri at this point. Maybe in this if there should be some sort of default configuration based on system properties, that would make it choose the caching or not caching path?

Reply:
Nice catch... the other solution would be to simply not support string/url/uri.  I can already see this being an issue when trying to lookup stream implementations as other input streams already advertise string/url etc as their input classes.  Maybe I should keep this class as only accepting `CogUri` so there is no confusion?

Reply:
I'm afraid it's not going to work in GeoServer if strings/URLs are not accepted, but I may be wrong (or it might be that we can build a CogUri there, just don't think it's possible to do so now, if memory serves me right every reader works off java own URL).